packages/contracts/contracts/examples/UnirepVoting.sol

```diff
@@ -0,0 +1,148 @@
+        for (uint i; i < numTeams; i++) {
+            projectData[i] = new uint256[](4);
+        }
+    }
+
+    // sign up users in this app
+    function userSignUp(
+        uint256[] calldata publicSignals,
+        uint256[8] calldata proof
+    ) public {
+        unirep.userSignUp(publicSignals, proof);
+    }
+
+    function joinProject(
+        uint256 projectID,
+        uint256[] memory publicSignals,
+        uint256[8] memory proof
+    ) public {
+        require(projectID < numTeams, 'Voteathon: invalid project iD');
+        EpochKeyVerifierHelper.EpochKeySignals memory signals = epochKeyHelper.verifyAndCheck(publicSignals, proof);
+
+        require(signals.revealNonce == true, 'Voteathon: should reveal nonce');
+        require(signals.nonce == 0, 'Voteathon: invalid nonce');
+        require(counts[projectID] < 10, 'Voteathon: maximum participants in a project');
+        participants[projectID].push(signals.epochKey);
+        // give user data if there is attestation before
+        uint48 epoch = unirep.attesterCurrentEpoch(uint160(address(this)));
+        require(epoch == 0, 'Voteathon: not join epoch');
+        uint256[] memory data = projectData[projectID];
+        for (uint256 i = 0; i < data.length; i++) {
+            unirep.attest(signals.epochKey, epoch, i, data[i]);
+        }
+        counts[projectID] += 1;
+    }
+
+    function vote(
+        uint256 projectID,
+        Option option,
+        uint256[] calldata publicSignals,
+        uint256[8] calldata proof
+    ) public {
+        require(projectID < numTeams, "projectID out of range");
+
+        EpochKeyVerifierHelper.EpochKeySignals memory signals = epochKeyHelper.verifyAndCheck(publicSignals, proof);
+
+        require(signals.revealNonce == true, "reveal nonce wrong");
+        require(signals.nonce == 1, "nonce wrong");
+
+        projectData[projectID][uint(option)] += 1;
+        
+        voted[signals.epochKey] += 1;
+        if (option == Option.UP) scores[projectID] += 1;
+        else if (option == Option.DOWN) scores[projectID] -= 1;
+
+        uint[] memory members = participants[projectID];
+        uint48 epoch = unirep.attesterCurrentEpoch(uint160(address(this)));
+        require(epoch == 0, 'Voteathon: not voting epoch');
+        for (uint256 i = 0; i < members.length; i++) {
+            unirep.attest(members[i], epoch, uint(option), 1);
+        }
+        
+        uint160 attesterId = uint160(address(this));
+        unirep.attesterStateTreeRootExists(attesterId, epoch, signals.stateTreeRoot);
+    }
+    
+    function claimPrize(
+        address receiver,
+        uint256[] calldata publicSignals,
+        uint256[8] calldata proof
+    ) public {
+        uint160 attesterId = uint160(address(this));
+        require(unirep.attesterCurrentEpoch(attesterId) > 0);
+        ReputationVerifierHelper.ReputationSignals memory signals = helper.verifyAndCheck(publicSignals, proof);
+        uint48 epoch = 0; // the voting epoch
```

Comment:
It should be the next epoch that the user transitions to. (`epoch > 0`)
There should be no data before user transition (`epoch=0`)
and we want to verify the latest proof (`epoch>0`)

Reply:
this epoch 0 is used to check: unirep.attesterStateTreeRootExists(attesterId, epoch, stateTreeRoot);

this is what the epoch that vote function is called: 
```
        require(epoch == 0, 'not voting epoch');
        for (uint256 i = 0; i < members.length; i++) {
            unirep.attest(members[i], epoch, uint(option), 1);
        }
```

Reply:
ah then it should still check `epoch>0`
```solidity
require(epoch > 0, 'should be after voting epoch');
uint256 stateTreeRoot = signals.stateTreeRoot;
unirep.attesterStateTreeRootExists(attesterId, epoch, stateTreeRoot);
```

Reply:
i think this is what you mean?

1) in `claim_prize`: check the epoch in the reputation proof for should > 0 
2) in `vote`: the epoch in the epoch key proof epoch should  = 0

revised to include them